Mono.Debugging/Mono.Debugging.Client/BreakpointStore.cs

```diff
@@ -572,14 +615,31 @@ public void FileRenamed (string oldPath, string newPath)
 		{
 			if (IsReadOnly)
 				return;
+
 			foreach (var bp in GetBreakpointsAtFile (oldPath)) {
 				Remove (bp);
 				bp.SetFileName (newPath);
 				Add (bp);
 			}
 		}
+
+		ImmutableArray<BreakEvent> SetBreakpoints (ImmutableArray<BreakEvent> newBreakpoints)
+		{
+			System.Diagnostics.Debug.Assert (System.Threading.Monitor.IsEntered (breakpointLock), "SetBreakpoints must be called during a lock");
+
+			var oldEvents = breakpoints;
+			breakpoints = newBreakpoints;
+			return oldEvents;
+		}
+
+		ImmutableArray<BreakEvent> InternalGetBreakpoints ()
+		{
+			lock (breakpointLock) {
+				return breakpoints;
+			}
```

Comment:
Locking access to an ImmutableArray does not look right to me. It is a thread safe type. Locking a non-thread safe collection such as a List would make sense.

Reply:
Agreed. That lock is unnecessary.

Reply:
A bit more explanation. Calling InternalGetBreakpoints will return a breakpoint list. This list can be the one that exists before calling InternalGetBreakpoints, or a new one if new breakpoints are added or removed during the execution of InternalGetBreakpoints. The lock doesn't really have any effect on which list will be returned, since the change could be done at any time, before the lock is taken or after. Also, it is safe to access the breakpoints properties without a lock, since the reference assignment is atomic.

Reply:
this "Also, it is safe to access the breakpoints properties without a lock, since the reference assignment is atomic." was what I was curious about and thought that might have been but when I discussed with Marius to check my understanding he did not think it was.

Reply:
Reference assignments are certainly atomic. The lock is not necessary.

Reply:
`breakpoints` itself is thread safe, what I wanted to validate by asking Marius was if the read of the breakpoints field would be guaranteed to be atomic because the field value could change from any other thread whilst reading. Perhaps we were not on the same page at the time, but that is the reason for the locks on reading.

Reply:
Ok, the problem is that ImmutableArray is not a reference type, it is a struct, and value type assignments are not always thread safe. Let me find out. The lock might be necessary after all!

Reply:
ImmutableArray assignment is most likely not atomic, hence why [InterlockedExchange](https://source.dot.net/#System.Collections.Immutable/System/Collections/Immutable/ImmutableInterlocked.cs,110) exists

Reply:
Which makes use of a nifty trick, and that is atomically swapping the underlying arrays of the structs.

Reply:
Given that ImmutableArray<T> has one single field that is an array reference, I'm pretty sure assignment is atomic. InterlockedExchange does more than an assignment, it reads the value and assigns the new value all in an atomic operation.